tests/unit/scripting.tcl

```diff
@@ -1578,3 +1578,69 @@ start_server {tags {"scripting"}} {
     }
 }
 
+# Cluster scripting tests
+
+# make sure the test infra won't use SELECT
+set old_singledb $::singledb
+set ::singledb 1
+
+start_server {overrides {cluster-enabled yes} tags {external:skip cluster}} {
+    r 0 cluster addslotsrange 0 16383
+    wait_for_condition 50 100 {
+        [csi 0 cluster_state] eq "ok"
+    } else {
+        fail "Cluster never became 'ok'"
+    }
+
+    test {Eval scripts with shebangs and functions default to no cross slots} {
+        # Test that scripts with shebang block cross slot operations
+        assert_error "ERR Script attempted to access keys that do not hash to the same slot*" {
+            r 0 eval {#!lua
+                redis.call('set', 'foo', 'bar')
+                redis.call('set', 'bar', 'foo')
+                return 'OK'
+            } 0}
+
+        # Test the functions by default block cross slot operations
+        r 0 function load REPLACE {#!lua name=crossslot
+            local function test_cross_slot(keys, args)
+                redis.call('set', 'foo', 'bar')
+                redis.call('set', 'bar', 'foo')
+                return 'OK'
+            end
+
+            redis.register_function('test_cross_slot', test_cross_slot)}
+        assert_error "ERR Script attempted to access keys that do not hash to the same slot*" {r FCALL test_cross_slot 0}
+    }
+
+    test {Cross slot commands are allowed by default for eval scripts and with allow-cross-slot-keys flag} {
+        # Old style lua scripts are allowed to access cross slot operations
+        r 0 eval "redis.call('set', 'foo', 'bar'); redis.call('set', 'bar', 'foo')" 0
+
+        # scripts with allow-cross-slot-keys flag are allowed
+        r 0 eval {#!lua flags=allow-cross-slot-keys
+            redis.call('set', 'foo', 'bar'); redis.call('set', 'bar', 'foo')
+        } 0
+
+        # Functions with allow-cross-slot-keys flag are allowed
+        r 0 function load REPLACE {#!lua name=crossslot
+            local function test_cross_slot(keys, args)
+                redis.call('set', 'foo', 'bar')
+                redis.call('set', 'bar', 'foo')
+                return 'OK'
+            end
+
+            redis.register_function{function_name='test_cross_slot', callback=test_cross_slot, flags={ 'allow-cross-slot-keys' }}}
+        r FCALL test_cross_slot 0
+    }
+    
+    test {Cross slot commands are also blocked if they disagree with pre-declared keys} {
+        assert_error "ERR Script attempted to access keys that do not hash to the same slot*" {
+            r 0 eval {#!lua
+                redis.call('set', 'foo', 'bar')
+                return 'OK'
+            } 1 bar}
+    }
```

Comment:
ohh, now i understand why you've put it in the client struct.
well, i'm not certain we need that feature.

maybe a better feature in that sense would be to block access to any key that's not declared, but that might be too much work...
if we go this way (current code), we should at least improve it to also cover reject (all) key access if the script was executed without specifying **any** key arguments.
WDYT?

Reply:
@oranagra I think we discussed that briefly in the past. I think it makes sense to not tolerate any form of cross key when this is enabled, and it's also completely orthogonal to undeclared key access.

Reply:
@yossigo i'm not sure i understand which solution you're promoting.
are you in favor of the current code (matches the slot of the actual keys being accessed and the one being declared?
or separating these two to different features?

Reply:
thinking of this further i can think of three different types of violations involving the declared keys (assuming the actual keys all belong to the same slot):
1. the declared key and the actual keys match different slots but both belong to the current node.
2. the declared key and actual keys match different slots. the declared one is local, and the actual keys are non-local.
3. the declared key and actual keys match different slots. the declared one is non-local, and the actual keys are local. 

outcome:
* in case [3], the command will get a MOVED error.
* in case [2], the commands inside the script will get rejected.
* in case [1], in theory there's real issue (effects replication will generate a MULTI-EXEC and all the keys it refer belong to the same slot).

arguably, in case [1], that command could be broken when resharding happens, but maybe there's some smart client that automatically sets a fake key name in the command that corresponds to the first slot of some node, just in order to make sure the command gets there, and has a different logic inside the script?
i.e. i can see a client doing such trick on scripts that use SCAN in order to target a different node each time, and the script itself accesses the local keys.

come to think of it, this feature will break that use case even if we don't match the slot of the declared key, so such a script will have to set the flag we created.

not sure i have any argument by now, but decided to post this anyway to promote discussion.

Reply:
> if we go this way (current code), we should at least improve it to also cover reject (all) key access if the script was executed without specifying any key arguments.

To reiterate to make sure I understand, if a script is executed without declaring any keys the expectation it won't be allowed to access any keys? It is possible a script could be smart enough to know how to operate on a given slot, and the slot might be passed in as a declared argument. 

> maybe a better feature in that sense would be to block access to any key that's not declared, but that might be too much work...

I agree with this in the long term, but that is more work to implement.

Reply:
> > if we go this way (current code), we should at least improve it to also cover reject (all) key access if the script was executed without specifying any key arguments.
> 
> To reiterate to make sure I understand, if a script is executed without declaring any keys the expectation it won't be allowed to access any keys? It is possible a script could be smart enough to know how to operate on a given slot, and the slot might be passed in as a declared argument.

well, such a script will be able to use the new flag we're adding.
i suppose the feature i'm suggesting here is a little bit orthogonal to the main topic of this PR.
it's just a minimal effort to have a (semi) default behavior that forces users to play nice.

come to think of it, with the current form of this PR you had to fix one bad test, i wonder how many bad tests you'll have to fix if you implement that suggestion.

i'm willing to drop this feature.. but i suppose that if we think it may be desired, now is the right time to add it...
although, arguably, it deserves a flag of it's own?

Reply:
I wasn't really able to convince myself one way or the other. I'm hesitant to add something so late though, but you are right that "now" is the time.

On one side I've seen folks use a pattern like `EVAL "whatever" 1 {fakekey} real-arg` for cases they don't know the key ahead of time, but they know the slot (or tag). In this case, they still use the keys in order to do cluster routing. 

On the other hand, it seems weird to force key declaration for non-clustered scripts. We could potentially allow functions to declare ways to define keys, but eval scripts don't really have a mitigation. So maybe it should only apply for cluster_enabled?

Reply:
for the record, we discussed that in the core team meeting and decided to leave that out